### PR TITLE
Define helper functions `setConstantDirichletBCLo` and `setConstantDirichletBCHi`

### PR DESCRIPTION
**Pull request by chongchonghe**
### Description

This PR introduces generic helper functions to encapsulate the logic for applying constant Dirichlet boundary conditions, thereby reducing code duplication and reduce chance of make mistakes in problem-specific implementations.

* **New Helper Functions for Boundary Conditions**: Introduced two templated helper functions for cell-centered variables, `setConstantDirichletBCLo` and `setConstantDirichletBCHi`, and two for face-centered variables, `setConstantDirichletBCFaceVarLo` and `setConstantDirichletBCFaceVarHi`, in `src/simulation.hpp` to standardize and simplify the application of constant Dirichlet boundary conditions on lower and upper boundaries, respectively.
* **Refactored Problem Boundary Conditions**: Updated problems to utilize the newly defined helper functions for setting custom constant boundary conditions.
* Also corrected some (non-harmful) errors. Many test problems uses `i >= hi[0]` to check for upper boundary cells, but it should be `i > hi[0]` because both lo and hi are inclusive, i.e. lo = [0, 0, 0] and hi = [7, 7, 7] for a 8^3 box. 

I also found and corrected a mistake (I believe) in 'RadTube' test: in `setCustomBoundaryConditions`, 'Erad' was set to `a T_lo^4` but 'Frad' was copied from the last real cell. This is not good because it could potential cause a violation of `Frad <= c Erad`. I changed it to `Frad = 0`. Similarly, the hydro states was falsely set to `v1 =` last real cell and `Egas = const.`, which could cause `Ekin > Etot`; I changed it to `v1 = 0`. 

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [ ] I have added a description (see above).
- [ ] I have added a link to any related issues (if applicable; see above).
- [ ] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

### CODE REVIEW

**Comment by chongchonghe** [comment]
@BenWibking Can you review what I have implemented so far before I move on to refactor all test problems?

**Comment by BenWibking** [comment]
Is this needed for either the tall box or INCITE sims?

**Comment by chongchonghe** [comment]
@BenWibking Not required, but helpful for my tall box runs. I will also implement a helper function for diode BC in this PR. A unified interface for custom BC would be helpful in avoiding mistakes.

**Comment by chongchonghe** [comment]
The reason I'm doing this is because I've seen inconsistent implementation of custom boundary conditions in the problem generators. 
For example, HydroSMS is doing `if (i < lo[0]) {...} else if (i >= hi[0]) {...}`:
https://github.com/quokka-astro/quokka/blob/e8193fba819e7dbc36527bc26b2b1118be1600ae/src/problems/HydroSMS/testHydroSMS.cpp#L122 

but RadTube is using `if (i < lo[0]) {...} else if (i > hi[0]) {...}`:
https://github.com/quokka-astro/quokka/blob/e8193fba819e7dbc36527bc26b2b1118be1600ae/src/problems/RadTube/testRadTube.cpp#L240

This PR adds a unified interface to avoid mistakes like this.

**Comment by BenWibking** [comment]
Ah, it looks like I screwed it up. I agree that's a good reason.

**Review by BenWibking — COMMENTED** [review]
I think this is a good design.

You could also fix my earlier inconsistency and call everything `lo` and `hi` instead of `left` and `right`.

**Comment by chongchonghe** [comment]
OK. 

BTW, those inconsistent `else if (i >= hi[0])` won't cause any bug as long as `setCustomBC` is called before fill in reflecting/periodic BC, because those ghost cells that shouldn't be filled will be overwritten later.

**Comment by chongchonghe** [comment]
All these are problematic. It should be `i > hi[0]` because both lo and hi are inclusive, i.e. `lo = [0, 0, 0]` and `hi = [7, 7, 7]` for a 8^3 box. Although, as explained above, they won't cause bugs. 

CleanShot 2026-01-16 at 15 11 25

File: src/simulation.hpp [hunk 1]
```diff
@@ -2369,6 +2413,190 @@ AMRSimulation<problem_t>::setCustomBoundaryConditionsFaceVar(const amrex::IntVec
...
+		cellIdx = i;
+	} else if constexpr (dir == 1) {
+		cellIdx = j;
+	} else if constexpr (dir == 2) {
+		cellIdx = k;
+	}
+
+	// Check if on upper boundary in the specified dimension and set values
+	// Note that both lo and hi are inclusive bounds, i.e. for a dimension with 8 cells, the bounds are 0 and 7.
+	if (cellIdx > hi[dir]) {
```
[the remark this answers is not in the document]
Comment (chongchonghe) [reply src/simulation.hpp]: the current implementation is correct and previous implementation has mistakes. Both `lo` and `hi` are inclusive.

**Comment by chongchonghe** [comment]
@BenWibking @markkrumholz Can you review this PR? It's ready to merge.

**Review by markkrumholz — APPROVED** [review]
